services/wallet/transfer/controller.go

```diff
@@ -298,3 +293,79 @@ func (c *Controller) GetCachedBalances(ctx context.Context, chainID uint64, addr
 
 	return blocksToViews(result), nil
 }
+
+var blockDuration int64 = 12
+
+func inferBlockCountAndStartTimestamp(lastTimestamp int64, secondsToNow int64) (historyBlockCount *big.Int, startTimestamp int64) {
+	startTimestamp = lastTimestamp - secondsToNow
+	blockCount := secondsToNow / blockDuration
+	historyBlockCount = big.NewInt(blockCount)
+	return historyBlockCount, startTimestamp
+}
+
+type BalanceState struct {
+	Value     big.Int `json:"value"`
+	Timestamp uint64  `json:"time"`
+}
+
+var BalanceHistorySecondsToNowFromInitialBlock int64 = -1
+
+// GetBalanceHistoryOnChain expect a time precision of +/- block time (12s)
+func (c *Controller) GetBalanceHistoryOnChain(ctx context.Context, chainID uint64, address common.Address, secondsToNow int64, sampleCount int) ([]BalanceState, error) {
+	chainClient, err := chain.NewClient(c.rpcClient, chainID)
+	if err != nil {
+		return nil, err
+	}
+
+	lastBlock, err := chainClient.BlockByNumber(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	lastBlockNumber := lastBlock.Number()
+	lastTimestamp := int64(lastBlock.Time())
+	if secondsToNow > lastTimestamp {
+		secondsToNow = lastTimestamp
+	}
+
+	firstBlockNo := big.NewInt(0)
+	historyBlockCount, startTimestamp := inferBlockCountAndStartTimestamp(lastTimestamp, secondsToNow)
+	if secondsToNow < 0 || historyBlockCount.Cmp(lastBlockNumber) > 0 {
+		initialBlock, err := chainClient.BlockByNumber(ctx, big.NewInt(0))
+		if err != nil {
+			return nil, err
+		}
+		firstBlockNo.Set(initialBlock.Number())
+		startTimestamp = int64(initialBlock.Time())
+		historyBlockCount.Sub(lastBlockNumber, firstBlockNo)
+	} else {
+		firstBlockNo.Sub(lastBlockNumber, historyBlockCount)
+	}
+
+	blocksInStep := historyBlockCount.Int64() / int64(sampleCount)
+	stepDuration := blocksInStep * blockDuration
+
+	points := make([]BalanceState, 0)
+
+	balanceCache := newBalanceCache()
```

Comment:
I might be wrong but this doesn't save anything into the db, only memory, and this is quite intensive in terms of infura call, due to our limitation, could we not save the result in db so that we do it once per block

Reply:
I agree. However, due to out-of-time DB caching won't make it in this PR but the functionality can be already tested. Will come back with a follow-up task for that.